src/Parsing/Impl/FStringParser.cs

```diff
@@ -0,0 +1,213 @@
+
+            // Called on '{'
+            Debug.Assert(CurrentChar() != '{', "Open brace expected");
+            NextChar();
+
+            int startExprPosition = _position;
+            int startExprLineNumber = _currentLineNumber;
+            int startExprColNumber = _currentColNumber;
+            Stack<char> nestedParens = new Stack<char>();
+            var errorReported = false;
+            while (!(EndOfFString())) {
+                char ch = CurrentChar();
+                if (nestedParens.Count == 0 && (ch == '}' || ch == '!' || ch == ':')) {
+                    // check that it's not a != comparison
+                    if (ch != '!' || !IsNext(notEqualStringSpan)) {
+                        break;
+                    }
+                }
+                errorReported = false;
+
+                if (ch == '#') {
+                    /* Error: can't include a comment character, inside parens
+                       or not. */
+                    ReportSyntaxError("f-string expression part cannot include '#'");
+                    errorReported = true;
+                } else if (ch == ')' || ch == '}') {
+                    _buffer.Append(ch);
+                    if (nestedParens.Pop() != ch) {
+                        ReportSyntaxError("");
+                        errorReported = true;
+                    }
+                } else if (ch == '(' || ch == '{') {
+                    nestedParens.Push(ch);
+                    _buffer.Append(ch);
+                } else {
+                    _buffer.Append(ch);
+                }
+
+                NextChar();
+            }
+            if (EndOfFString()) {
+                if (nestedParens.Count > 0) {
+                    ReportSyntaxError("");
+                } else {
+                    ReportSyntaxError("");
+                }
+                return new ErrorExpression(_buffer.ToString(), null);
+            }
+            Expression fStringExpression;
+            if (CurrentChar() == '}' || CurrentChar() == '!' || CurrentChar() == ':') {
+
+                string subExprStr = _buffer.ToString();
+                _buffer.Clear();
+                fStringExpression = CreateExpression(subExprStr,
+                        startExprPosition, startExprLineNumber, startExprColNumber);
+
+                if (CurrentChar() == '}') {
+                    return fStringExpression;
+                }
+            } else {
+                // try to recover
+                //return ErrorExpression();
+                return null;
+            }
+            var conversion = ' ';
+            Expression formatSpecifier = null;
+            if (CurrentChar() == '!') {
+                Read('!');
+                conversion = NextChar();
+                if (!(conversion == 's' || conversion == 'r' || conversion == 'a')) {
+                    ReportSyntaxError($"f-string: invalid conversion character: {conversion} expected 's', 'r', or 'a'");
```

Comment:
Since we have https://github.com/Microsoft/python-language-server/issues/647 can you please relocate to resources strings at least in the new code.